extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/TaskRunner.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.kafka.connect.source.SourceTaskContext;
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
+
+class TaskRunner {
+    private static final ILogger LOGGER = Logger.getLogger(TaskRunner.class);
+    private final SourceConnector connector;
+    private final ReentrantLock taskLifecycleLock = new ReentrantLock();
+    private final Map<Map<String, ?>, Map<String, ?>> partitionsToOffset;
+    private volatile boolean running;
+    private volatile boolean reconfigurationRequested;
+    private SourceTask task;
+
+    TaskRunner(SourceConnector connector, Map<Map<String, ?>, Map<String, ?>> partitionsToOffset) {
+        this.connector = connector;
+        this.partitionsToOffset = partitionsToOffset;
+    }
+
+    List<SourceRecord> poll() {
+        restartTaskIfNeeded();
+        return doPoll();
+    }
+
+    void stop() {
+        LOGGER.info("Stopping task if running");
+        try {
+            taskLifecycleLock.lock();
+            if (running) {
+                LOGGER.fine("Stopping task");
+                task.stop();
+                LOGGER.fine("Task stopped");
+            }
+        } finally {
+            running = false;
+            task = null;
+            taskLifecycleLock.unlock();
+        }
+    }
+
+    void requestReconfiguration() {
+        LOGGER.info("Task reconfiguration requested");
+        reconfigurationRequested = true;
+    }
+
+    private List<SourceRecord> doPoll() {
+        try {
+            return task.poll();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw rethrow(e);
+        }
+    }
+
+    private void restartTaskIfNeeded() {
+        if (reconfigurationRequested) {
+            reconfigurationRequested = false;
+            stop();
+        }
+        start();
+    }
+
+    private void start() {
+        LOGGER.info("Starting tasks if the previous not running");
+        try {
+            taskLifecycleLock.lock();
+            if (!running) {
+                LOGGER.info("Starting tasks");
+
+                List<Map<String, String>> taskConfigs = connector.taskConfigs(1);
+                if (!taskConfigs.isEmpty()) {
+                    Map<String, String> taskConfig = taskConfigs.get(0);
+                    this.task = newInstance(connector.taskClass().asSubclass(SourceTask.class));
+                    LOGGER.info("Initializing task: " + task);
+                    task.initialize(new JetSourceTaskContext(taskConfig));
+                    LOGGER.info("Starting task: " + task + " with task config: " + taskConfig);
+                    task.start(taskConfig);
+                    running = true;
+                } else {
+                    LOGGER.info("No task configs!");
```

Comment:
Isn't it an exception?

Reply:
As mentioned in the PR description: `it's required for connectors that perform asynchronous initialisation`. E.g. Kafka Connect JDBC Connector scans DB asynchronously for newly created tables. It can happen that jobs are started before the initial scan of existing tables finished which leads to no task configs. After some time, in the next iteration of the `KafkaConnectSource#fillBuffer` we will got a task config.